comp/core/tagger/taggerimpl/tagstore/entity_tags.go

```diff
@@ -147,3 +200,160 @@ func (e *EntityTags) shouldRemove() bool {
+	sources := make([]string, 0, len(e.sourceTags))
+	for source := range e.sourceTags {
+		sources = append(sources, source)
+	}
+	return sources
+}
+
+func (e *EntityTagsWithMultipleSources) deleteSource(source string, expiryDate time.Time) {
+	tags, ok := e.sourceTags[source]
+	if !ok {
+		return
+	}
+
+	tags.expiryDate = expiryDate
+	e.sourceTags[source] = tags
+}
+
+// EntityTagsWithSingleSource holds the tag information for a given entity that
+// can only be created from a single source. It is not thread-safe, so should
+// not be shared outside of the store. Usage inside the store is safe since it
+// relies on a global lock.
+type EntityTagsWithSingleSource struct {
+	entityID           string
+	source             string
+	expiryDate         time.Time
+	standardTags       []string
+	cachedAll          tagset.HashedTags // Low + orchestrator + high
+	cachedOrchestrator tagset.HashedTags // Low + orchestrator (subslice of cachedAll)
+	cachedLow          tagset.HashedTags // Sub-slice of cachedAll
+	isExpired          bool
+}
+
+func newEntityTagsWithSingleSource(entityID string, source string) *EntityTagsWithSingleSource {
+	return &EntityTagsWithSingleSource{
+		entityID: entityID,
+		source:   source,
+	}
+}
+
+func (e *EntityTagsWithSingleSource) toEntity() types.Entity {
+	cachedAll := e.cachedAll.Get()
+	cachedOrchestrator := e.cachedOrchestrator.Get()
+	cachedLow := e.cachedLow.Get()
+
+	return types.Entity{
+		ID:           e.entityID,
+		StandardTags: e.getStandard(),
+		// cachedAll contains low, orchestrator and high cardinality tags, in this order.
+		// cachedOrchestrator and cachedLow are subslices of cachedAll, starting at index 0.
+		HighCardinalityTags:         cachedAll[len(cachedOrchestrator):],
+		OrchestratorCardinalityTags: cachedOrchestrator[len(cachedLow):],
+		LowCardinalityTags:          cachedLow,
+	}
+}
+
+func (e *EntityTagsWithSingleSource) getStandard() []string {
+	return e.standardTags
+}
+
+func (e *EntityTagsWithSingleSource) getHashedTags(cardinality types.TagCardinality) tagset.HashedTags {
+	switch cardinality {
+	case types.HighCardinality:
+		return e.cachedAll
+	case types.OrchestratorCardinality:
+		return e.cachedOrchestrator
+	default:
+		return e.cachedLow
+	}
+}
+
+func (e *EntityTagsWithSingleSource) tagsForSource(source string) *sourceTags {
+	if source != e.source {
+		return nil
+	}
+
+	return &sourceTags{
+		lowCardTags:          e.cachedLow.Get(),
+		orchestratorCardTags: e.cachedAll.Slice(e.cachedLow.Len(), e.cachedOrchestrator.Len()).Get(),
+		highCardTags:         e.cachedAll.Slice(e.cachedOrchestrator.Len(), e.cachedAll.Len()).Get(),
+		standardTags:         e.standardTags,
+		expiryDate:           e.expiryDate,
+	}
+}
+
+func (e *EntityTagsWithSingleSource) tagsBySource() map[string][]string {
+	return map[string][]string{e.source: e.cachedAll.Get()}
+}
+
+func (e *EntityTagsWithSingleSource) setTagsForSource(source string, tags sourceTags) {
+	if source != e.source {
+		return
```

Comment:
should we emit some log (or even a panic) here to indicate that something doesn't look right.

Reply:
I added this in the new commit 👍